src/Database/Esqueleto/Internal/Internal.hs

```diff
@@ -2884,3 +2885,27 @@ insertSelect = void . insertSelectCount
 insertSelectCount :: (MonadIO m, PersistEntity a) =>
   SqlQuery (SqlExpr (Insertion a)) -> SqlWriteT m Int64
 insertSelectCount = rawEsqueleto INSERT_INTO . fmap EInsertFinal
+
+-- | Avoid N+1 queries and join entities into a map structure
+-- @
+-- getFoosAndNestedBarsFromParent :: ParentId -> (Map (Key Foo) (Foo, [Entity Bar]))
+-- getFoosAndNestedBarsFromParent parentId = 'fmap' associateJoin $ 'select' $
+-- 'from' $ \\(foo `'LeftOuterJoin`` bar) -> do
+--   'on' (bar '^.' BarFooId '==.' foo '^.' FooId)
```

Comment:
The `bar` access here should use `?.` because a `LeftOuterJoin` implies nullability on the right hand side. That this isn't enforced is a bug (see #150). The type of this would then be `Map (Key Foo) (Foo, [Maybe (Entity Bar)])`.